### PR TITLE
fix: tabbable sometimes incorrectly includes children of a web component

### PR DESCRIPTION
**Pull request by benjamin-t-frost**
## Use Case
A browser will skip tab-targeting an element if it is the only child of a web component which has a tabindex of -1.

## Example

```html
<my-input tabindex="-1">
  <input id="input" type="text" />
</my-nput>
```

## Expected behaviour
`tabbable()` should return an empty list for the example above

## Actual behaviour
`tabbable()` returns a list which includes the input depicted above

## Impact
This bug can cause a focus-trap to behave erratically when nearing the end of focusable elements.

<details>
<summary>PR Checklist</summary>
<br/>

__Please leave this checklist in your PR.__

- Source changes maintain stated browser compatibility.
- Issue being fixed is referenced.
- Unit test coverage added/updated.
- E2E test coverage added/updated.
- Typings added/updated.
- Changes do not break SSR:
  - Careful to test `typeof document/window !== 'undefined'` before using it in code that gets executed on load.
- README updated (API changes, instructions, etc.).
- Changes to dependencies explained.
- Changeset added (run `yarn changeset` locally to add one, and follow the prompts).
  - EXCEPTION: A Changeset is not required if the change does not affect any of the source files that produce the package bundle. For example, demo changes, tooling changes, test updates, or a new dev-only dependency to run tests more efficiently should not have a Changeset since it will not affect package consumers.

</details>

### CODE REVIEW

**Comment by stefcameron** [comment]
👀

File: src/index.js [hunk 1]
```diff
@@ -415,6 +420,15 @@ const isNodeMatchingSelectorTabbable = function (options, node) {
   return true;
 };
 
+const isValidShadowCandidatesTabbable = function (node, nestedCandidates) {
```
Comment (stefcameron) [inline src/index.js:423]: So here, `node` is expected to be the shadow host node, and `nestedCandidates` are all the candidate tabbable/focusable (already vetted as such) nodes that were found inside of it?

If so, could we clarify the expectation by renaming `node` to `shadowHostNode`?

File: test/fixtures/shadow-dom-untabbable.html [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<test-shadow tabindex="-1">
+  <template shadowroot="open">
+    <div id="container">
+      <input id="input" type="text" />
+    </div>
+  </template>
+</test-shadow>
```
Comment (stefcameron) [inline test/fixtures/shadow-dom-untabbable.html:7]: I appreciate the new test, thank you! Could we also have another test that shows when the `nestedCandidates.length !== -1` case, that is, a test that checks the opposite of what this new test is checking?

Based on your explanation in the code:
```
  // If a custom element with a tabindex of -1 contains a single otherwise tabbable
  // element, browsers will not allow tab targeting said child.
  return nestedCandidates.length !== 1;
```
I'm getting the impression that if I changed your test to:
```
<test-shadow tabindex="-1">
  <template shadowroot="open">
    <div id="container">
      <input id="input" type="text" />
      <input id="input2" type="text" />
    </div>
  </template>
</test-shadow>
```
that now, both inputs would be tabbable? Seems a bit strange, so I wanted to double-check. Having a passing test that checks for the opposite is always a good reminder of all the corner cases.

**Review by stefcameron — CHANGES_REQUESTED** [review]
A very proactive PR, ticking all the boxes. 😄 Thank you! I do have a couple of requests/concerns, though.

File: src/index.js [hunk 3]
```diff
@@ -415,6 +420,15 @@ const isNodeMatchingSelectorTabbable = function (options, node) {
...
+  if (node.getAttribute('tabindex') !== '-1') {
```
[the remark this answers is not in the document]
Comment (benjamin-t-frost) [reply src/index.js]: After testing, I've changed this to filter out anything which is explicitly negative. `getTabindex()` returns implicit `-1`s and causes things to be filtered out which should not.

File: test/fixtures/shadow-dom-untabbable.html (hunk 2 above)
Comment (stefcameron) [reply test/fixtures/shadow-dom-untabbable.html:7]: Definitely! Thanks for checking it out.

**Review by stefcameron — APPROVED** [review]
Thanks for the updates, and for tweaking the comment in the code about what is and is not tabbable to be more generic so there's no confusion in the future about shadows with 1 vs multiple children.

Looks good!

**Comment by stefcameron** [comment]
`focus-trap@6.9.3` is getting published now. :)
